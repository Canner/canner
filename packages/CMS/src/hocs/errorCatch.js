// @flow

import * as React from 'react';
import {Alert} from 'antd';

import type {HOCProps} from './types';

type State = {
  error: any,
  errorInfo: any
};

export default function errorCatch(Com: React.ComponentType<*>) {
  return class ComponentErrorCatch extends React.Component<HOCProps, State> {
    state = {
      error: false,
      errorInfo: {componentStack: null}
    }

    componentDidCatch(e: Error, info: Object) {
      this.setState({
        error: e,
        errorInfo: info
      });
    }

    render() {
      const {error, errorInfo} = this.state;
      if (error) {
        return <Alert
          message="Something went wrong."
          description={(
            <details>
              {error.toString()}
              <br />
              {errorInfo.componentStack}
            </details>
          )}
          type="error"
          closable
        />
      }
      return <Com {...this.props} />
    }
  };
}