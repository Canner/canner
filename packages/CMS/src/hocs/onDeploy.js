// @flow

import * as React from 'react';
import RefId from 'canner-ref-id';
import {Map, List} from 'immutable';
import type {HOCProps} from './types';

export default function withOndeploy(Com: React.ComponentType<*>) {
  return class ComponentWithOnDeploy extends React.Component<HOCProps> {
    key: string;
    id: ?string;

    constructor(props: HOCProps) {
      super(props);
      const {pattern, refId, rootValue} = props;
      const {key, id} = splitRefId({
        refId,
        rootValue,
        pattern
      });
      this.key = key;
      this.id = id;
    }

    removeOnDeploy = (arg1: string, callbackId: string) => {
      const {removeOnDeploy} = this.props;
      if (callbackId) {
        return removeOnDeploy(arg1, callbackId);
      } else {
        return removeOnDeploy(this.key, arg1);
      }
    }

    onDeploy = (arg1: string | Function, callback: Function): string => {
      const {onDeploy, refId} = this.props;
      if (typeof arg1 === 'string') {
        return onDeploy(arg1, callback);
      } else {
        // first arguments is a function
        return onDeploy(this.key, result => {
          let restPathArr = refId.getPathArr();
          // if (this.id) {
          //   restPathArr = restPathArr.slice(2);
          // } else {
            restPathArr = restPathArr.slice(1);
          // }
          const {paths, value} = getValueAndPaths(result.data, restPathArr);
          return {
            ...result,
            // $FlowFixMe
            data: result.data.setIn(paths, arg1(value))
          }
        });
      }
    }

    render() {
      return <Com {...this.props}
        onDeploy={this.onDeploy}
        removeOnDeploy={this.removeOnDeploy}
      />
  }
  };
}


export function splitRefId({
  refId,
  rootValue,
  pattern
}: {
  refId: RefId,
  rootValue: any,
  pattern: string
}) {
  const [key, index] = refId.getPathArr();
  let id;
  if (pattern.startsWith('array')) {
    id = rootValue.getIn([key, index, 'id']);
  }
  return {
    key,
    id
  }
}

export function getValueAndPaths(value: Map<string, *>, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    let v = result.value;
    let paths = result.paths;
    if (Map.isMap(v)) {
      if (v.has('edges') && v.has('pageInfo')) {
        v = v.getIn(['edges', key, 'node']);
        paths = paths.concat(['edges', key, 'node']);
      } else {
        v = v.get(key);
        paths = paths.concat(key);
      }
    } else if (List.isList(v)) {
      v = v.get(key);
      paths = paths.concat(key);
    }
    return {
      value: v,
      paths
    }
  }, {
    value,
    paths: []
  });
}