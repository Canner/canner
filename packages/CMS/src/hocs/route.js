// @flow
import * as React from 'react';
import {Button, Icon, Spin} from 'antd';
import styled from 'styled-components';
import type {HOCProps} from './types';

const RENDER_CHILDREN = 0;
const RENDER_COMPONENT = 1;
const RENDER_NULL = 2;

const ButtonWrapper = styled.div`
  text-align: right;
`
type State = {
  deploying: boolean
}

export default function withRoute(Com: React.ComponentType<*>) {
  return class ComWithRoute extends React.Component<HOCProps, State> {
    state = {
      deploying: false
    };
  
    deploy = () => {
      const {refId, deploy, goTo, routes} = this.props;
      this.setState({
        deploying: true
      })
      deploy(refId.getPathArr()[0])
        .then(() => {
          this.setState({
            deploying: false
          });
          goTo(routes[0]);
        })
        .catch(() => {
          this.setState({
            deploying: false
          });
        });
    }

    reset = () => {
      const {refId, reset, goTo, routes} = this.props;
      reset(refId.getPathArr()[0])
        .then(() => goTo(routes[0]));
    }

    discard = () => {
      const {goTo, routes, params, reset, refId} = this.props;
      if (params.op === 'create') {
        reset(refId.getPathArr()[0]).then(() => goTo(routes.join('/')));
      } else {
        reset(refId.getPathArr()[0]).then(() => goTo(routes.slice(0, -1).join('/')));
      }
      
    }

    render() {
      const {deploying} = this.state;
      let {routes, pattern,  path, params, refId, renderChildren, hideButtons, uiParams} = this.props;
      const renderType = getRenderType({
        pattern,
        routes,
        path,
        params
      });
      const pathArrLength = refId.getPathArr().length;
      const routesLength = routes.length;
      const {op} = params;
      let renderKeys = uiParams && uiParams.updateKeys; // render all
      if (op === 'create') {
        renderKeys = uiParams && uiParams.createKeys;
      }
      return <Spin tip="deploying" spinning={deploying}>
        {
          // quick fix for route array's children
          // need to find a stable way to control route
          (renderType === RENDER_CHILDREN && pattern === 'array' && (routesLength === pathArrLength || (routesLength + 1 === pathArrLength && op === 'create'))) &&
            <Button onClick={this.discard} style={{marginBottom: 16}}>
              <Icon type="arrow-left" /> Back
            </Button>
        }
        {
          renderType === RENDER_CHILDREN && renderChildren(node => {
            return {
              hidden: renderKeys && renderKeys.indexOf(node.keyName) === -1,
              refId
            };
          })
        }
        {
          // quick fix for route array's children
          // need to find a stable way to control route
          (renderType === RENDER_CHILDREN  && pattern === 'array' && !hideButtons && (routesLength === pathArrLength || (routesLength + 1 === pathArrLength && op === 'create'))) &&
            <ButtonWrapper>
              <Button style={{marginRight: 16}} type="primary" onClick={this.deploy}>Confirm</Button>
              <Button onClick={this.reset}>Reset</Button>
            </ButtonWrapper>
        }
        {
          renderType === RENDER_COMPONENT && <Com {...this.props} />
        }
      </Spin>
    }
  }
}

function getRenderType({
  routes,
  path,
  pattern,
  params
}) {
  const paths = genPaths(path, pattern);
  const pathsLength = paths.length;
  const routesLength = routes.length;
  if (routes[0] !== paths[0]) {
    return RENDER_NULL;
  }
  const type = pattern.split('.').slice(-1)[0];
  const {op} = params;
  if (type === 'object') {
    if (routesLength === pathsLength && isCompleteContain(paths, routes)) {
      return RENDER_COMPONENT;
    }
    if (routesLength < pathsLength) {
      return RENDER_COMPONENT;
    }
    if (routesLength > pathsLength) {
      return RENDER_CHILDREN;
    }
    return RENDER_NULL;
  } else if (type === 'array') {
    if (routesLength === pathsLength && isCompleteContain(paths, routes)) {
      return RENDER_CHILDREN;
    }
    if (routesLength < pathsLength) {
      if (routesLength === pathsLength - 1 && op === 'create') {
        return RENDER_CHILDREN;
      }
      return RENDER_COMPONENT;
    }
    if (routesLength > pathsLength) {
      return RENDER_COMPONENT;
    }
    return RENDER_NULL;
  } else {
    if (routesLength === pathsLength && isCompleteContain(paths, routes)) {
      return RENDER_NULL;
    }
    if (routesLength < pathsLength) {
      return RENDER_COMPONENT;
    }
    if (routesLength > pathsLength) {
      return RENDER_COMPONENT;
    }
    return RENDER_NULL;
  }
}

export function isCompleteContain(paths: Array<string>, routes: Array<string>) {
  return paths.map((key, i) => routes[i] === key || key === '__ARRAY_INDEX__')
    .reduce((result, curr) => result && curr);
}

export function genPaths(path: string, pattern: string) {
  const patterns = pattern.split('.');
  const indexs = patterns.map((type, i) => type === 'array' ? i : -1)
    .filter(index => index !== -1);
  let paths = path.split('/')
    .map((route, i) => indexs.indexOf(i) === -1 ? route : [].concat(route, '__ARRAY_INDEX__'))
    .reduce((result, curr) => result.concat(curr), []);
  return paths;
}
