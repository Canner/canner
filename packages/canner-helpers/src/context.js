import React from 'react';

export const defaultContext =  {
  renderChildren: () => {throw new Error('There is no renderChildren method')},
  renderConfirmButton: () => {throw new Error('There is no renderConfirmButton method')},
  renderCancelButton: () => {throw new Error('There is no renderCancelButton method')},
  renderComponent: () => {throw new Error('There is no renderComponent method')},
  refId: null,
  routes: null,
};

function getContextFromWindow() {
  if (window) {
    window.__canner_item_contenxt__ = window.__canner_item_contenxt__ || React.createContext(defaultContext);
    return  window.__canner_item_contenxt__;
  }
  return {};
}

export default typeof window === 'undefined' ? React.createContext(defaultContext) : getContextFromWindow();

