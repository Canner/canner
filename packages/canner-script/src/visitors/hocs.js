/* eslint-disable no-param-reassign */
const visitors = {
  component: {
    enter(path) {
      // first layer
      path.node.hocs = ['withTitleAndDescription']; // the bottom HOC

      path.node.hocs.push('withRequest');
      path.node.hocs.push('relation');

      /**
       *  cache the popup data!
       */
      if (path.node.ui === 'table') {
        path.node.hocs.push('miniApp');
      }

      /**
       *  check whether this plugin need to be rendered,
       *  should be under the query
       *  because it needs the data to render the children of array!
       */
      path.node.hocs.push('withRoute');

      if (path.route.split('.').length === 1 && path.node.ui.startsWith('table')) {
        path.node.hocs.push('query');
      } else {
        path.node.hocs.push('withQuery');
      }

      /**
       *  route mini app must be place at the top of query
       *  because when route changes, it will get query.queryData to fetch the new Data
       */
      path.node.hocs.push('routeMiniApp');

      path.node.hocs.push('connectId'); // the top HOC
    },
  },
  layout: {
    enter(path) {
      path.node.hocs = ['containerRouter', 'context'];
      if (path.node.ui === 'row' || path.node.ui === 'col') {
        path.node.hocs = [];
      }
    },
  },
};

export default [visitors];
