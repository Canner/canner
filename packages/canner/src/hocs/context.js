// @flow
import * as React from 'react';

export const defaultContext = {
  query: {},
  dataChanged: {},
  fetch: () => {throw new Error('there is no fetch method');},
  subscribe: () => {throw new Error('there is no subscribe method');},
  request: () => {throw new Error('there is no request method');},
  deploy: () => {throw new Error('there is no deploy method');},
  reset: () => {throw new Error('there is not reset method');},
  onDeploy: () => {throw new Error('there is not onDeploy method');},
  removeOnDeploy: () => {throw new Error('there is not removeOnDeploy method');},
  updateQuery: () => {},
}

// $FlowFixMe: waiting for flow updata the libdef
export const HOCContext = React.createContext(defaultContext);
