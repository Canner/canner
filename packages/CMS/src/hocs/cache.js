// @flow
import * as React from 'react';
import {mutate as defaultMutate, ActionManager as DefaultAciontManager} from '../action';
import {isCompleteContain, genPaths} from './route';
import { isArray } from 'lodash';
import { OnDeployManager } from '../onDeployManager';
import type {Action, ActionType} from '../action/types';
import type {HOCProps} from './types';

type State = {
  [string]: *
}

export default function withCache(Com: React.ComponentType<*>, options: {
  mutate: typeof defaultMutate,
  ActionManager: DefaultAciontManager,
  onDeployManager: OnDeployManager
}) {
  const {mutate = defaultMutate, ActionManager = DefaultAciontManager} = options || {};
  return class ComWithCache extends React.Component<HOCProps, State> {
  actionManager: ?ActionManager;
  onDeployManager: OnDeployManager;
  subscribers: {
      [key: string]: Array<{id: string, callback: Function}>
    }
    subscribers = {};
    subscription: any;
    constructor(props: HOCProps) {
      super(props);
      const {routes, cacheActions, pattern, path} = this.props;
      if ((routes.length > 1 && isRoutesEndAtMe({routes, pattern, path})) ||
        cacheActions
      ) {
        this.actionManager = new ActionManager();
        this.onDeployManager = new OnDeployManager();
      }
      this.state = {
      };
    }

    addSubscriber = (key: string, id: string, callback: Function) => {
      const subscriber = {
        id,
        callback
      };
      if (this.subscribers[key]) {
        this.subscribers[key].push(subscriber);
      } else {
        this.subscribers[key] = [subscriber];
      }
    }

    unsubscribe = (key: string, subscriberId: string) => {
      this.subscribers[key] = this.subscribers[key].filter(subscriber => {
        return subscriber.id !== subscriberId;
      });
    }

    publish = (key: string, id?: string) => {
      // $FlowFixMe
      if (!this.actionManager) {
        return;
      }
      const data = this.state[key];
      if (!data) {
        return;
      }
      const actions = this.actionManager.getActions(key, id);
      const mutatedData = actions.reduce((result, action) => {
        return mutate(result, action);
      }, data);
      (this.subscribers[key] || []).forEach(subscribe => {
        subscribe.callback(mutatedData);
      });
    }

    fetch = (key: string) => {
      // the data will be mutated by cached actions
      const {fetch} = this.props;
      if (!this.actionManager) {
        return fetch(key);
      }
      const actions = this.actionManager.getActions(key);
      return fetch(key).then(data => {
        this.setState({
          [key]: data
        });
        this._subscribe(key);
        return actions.reduce((result, action) => {
          return mutate(result, action);
        }, data);
      });
    }

    _subscribe = (key: string) => {
      const {subscribe} = this.props;
      this.subscription = subscribe(key, (data) => {
        this.setState({
          [key]: data
        }, () => this.publish(key));
      });
    }

    _unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    request = (action: Array<Action<ActionType>> | Action<ActionType>): Promise<*> => {
      // use action manager cache the actions
      // update state.actions
      const {request} = this.props;
      if (!this.actionManager) {
        // $FlowFixMe
        return request(action);
      }
      let key, id;
      if (isArray(action)) {
        // $FlowFixMe
        action.forEach(ac => {
          // $FlowFixMe
          this.actionManager.addAction(ac);
        });
        key = action[0].payload.key;
        id = action[0].payload.id;
      } else {
        // $FlowFixMe
        this.actionManager.addAction(action);
        // $FlowFixMe
        key = action.payload.key;
        // $FlowFixMe
        id = action.payload.id;
      }
      this.publish(key, id);
      return Promise.resolve();
    }

    onDeploy = (key: string, callback: any) => {
      const {onDeploy} = this.props;
      if (!this.onDeployManager) {
        return onDeploy(key, callback);
      }
      return this.onDeployManager.registerCallback(key, callback);
    }

    removeOnDeploy = (key: string, callbackId: string) => {
      const {removeOnDeploy} = this.props;
      if (!this.onDeployManager) {
        return removeOnDeploy(key, callbackId);
      }
      return this.onDeployManager.unregisterCallback(key, callbackId);
    }

    _executeOnDeployCallback = (key: string, value: any) => {
      return this.onDeployManager.execute({
        key,
        value
      });
    }

    deploy = (key: string, id?: string): Promise<*> => {
      // request cached actions
      const {request, deploy, pattern} = this.props;
      if (!this.actionManager) {
        return deploy(key, id);
      }
      const originData = this.state[key];
      let actions = this.actionManager.getActions(key, id);
      const mutatedData = actions.reduce((result, action) => {
        return mutate(result, action);
      }, originData);
      const {error} = this._executeOnDeployCallback(key, mutatedData.get(key));
      if (error) {
        return Promise.reject();
      }
      // actions = actions.map(action => {
      //   const {key, value} = action.payload;
      //   hasError = hasError || error;
      //   action.payload.value = data;
      //   return action;
      // });

      // $FlowFixMe
      this.actionManager.removeActions(key, id);
      // $FlowFixMe
      request(actions);
      // if this cache is on the first layer,
      // it should call the deploy after request
      if (pattern.split('.').length === 1) {
        return deploy(key, id);
      }
      return Promise.resolve();
    }

    reset = (key: string, id?: string): Promise<*> => {
      // remove sepicfic cached actions in actionManager
      const {reset} = this.props;
      if (!this.actionManager) {
        return reset(key, id);
      }
      this.actionManager.removeActions(key, id);
      this.publish(key, id);
      return Promise.resolve();
    }

    subscribe = (key: string, callback: Function) => {
      const {subscribe} = this.props;
      if (!this.actionManager) {
        return subscribe(key, callback);
      }
      const id = genSubscriberId();
      this.addSubscriber(key, id, callback);
      return {
        unsubscribe: () => {
          this.unsubscribe(key, id);
        }
      }
    }

    updateQuery = (paths: Array<string>, args: Object) => {
      const {updateQuery} = this.props;
      const reWatch = updateQuery(paths, args);
      if (reWatch) {
        this._unsubscribe();
        this.fetch(paths[0]);
      }
    }

    render() {
      return (
        <Com
          {...this.props}
          fetch={this.fetch}
          request={this.request}
          deploy={this.deploy}
          reset={this.reset}
          subscribe={this.subscribe}
          updateQuery={this.updateQuery}
          onDeploy={this.onDeploy}
          removeOnDeploy={this.removeOnDeploy}
        />
      );
    }
  }
}

export function isRoutesEndAtMe({
  routes,
  path,
  pattern
}: {
  routes: Array<string>,
  path: string,
  pattern: string
}): boolean {
  const paths = genPaths(path, pattern);
  return (paths.length === routes.length && isCompleteContain(paths, routes));
}

export function genSubscriberId() {
  return Math.random().toString(36).substr(2, 7);
}