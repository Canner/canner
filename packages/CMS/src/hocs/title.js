// @flow
import * as React from 'react';
import {Tooltip, Icon, Alert} from 'antd';
import {HOCContext} from './context';
import {Context} from 'canner-helpers';
import styled from 'styled-components';
import {isEmpty} from 'lodash';
import type {HOCProps} from './types';

const Title = styled.div`
  font-size: 18px;
  font-weight: 400;
  &:after {
    content: ${props => props.required ? '"*"' : '""'};
    color: red;
  }
`;

// $FlowFixMe
export default function withTitleAndDescription(Com: React.ComponentType<*>) {
  return class ComponentWithTitleAndDescription extends React.Component<HOCProps> {
    render() {
      const {title, layout, description, hideTitle,
        fetch, subscribe, request, deploy, reset, query,
        renderChildren, renderComponent, renderConfirmButton, renderCancelButton,
        refId, routes, updateQuery, type, imageServiceConfig,
        onDeploy, removeOnDeploy, required
      } = this.props;
      return <HOCContext.Provider
        value={{
          fetch,
          subscribe,
          request,
          deploy,
          reset,
          query,
          updateQuery,
          onDeploy,
          removeOnDeploy
        }}
      >
        <Context.Provider value={{
          renderChildren,
          renderComponent,
          renderConfirmButton,
          renderCancelButton,
          refId,
          routes
        }}>
          {
            hideTitle && <Com {...this.props}/>
          }
          {
            (!hideTitle && layout === 'horizontal') && <div style={{
              display: 'flex',
              margin: '16px 0',
              alignItems: 'center'
            }}>
              <div style={{
                marginRight: 8,
                display: 'flex',
                flex: 1,
                flexDirection: 'row'
              }}>
                <Title required={required}>
                  {title}
                </Title>
                {
                  description && (
                    <Tooltip placement="top" title={description}>
                      <Icon type="info-circle-o" style={{marginLeft: 12, color: '#aaa'}}/>
                    </Tooltip>
                  )
                }
              </div>
              {
                (type === 'image' && isEmpty(imageServiceConfig)) && (
                  <Alert style={{margin: '16px 0'}} message="There is no storage config so you can't upload image. Checkout the storage section to know more" type="warning" />
                )
              }
              <div style={{
                flex: 2
              }}>
                <Com {...this.props}/>
              </div>
            </div>
          }
          {
            (!hideTitle && layout !== 'horizontal') && <div style={{
              margin: '16px 0 0'
            }}>
              <div style={{
                display: 'flex',
                alignItems: 'center'
              }}>
                <Title required={required}>
                  {title}
                </Title>
                {
                  description && (
                    <Tooltip placement="top" title={description}>
                      <Icon type="info-circle-o" style={{marginLeft: 12, color: '#aaa'}}/>
                    </Tooltip>
                  )
                }
                
              </div>
              {
                (type === 'image' && isEmpty(imageServiceConfig)) && (
                  <Alert style={{margin: '16px 0'}} message={<p>There is no storage config so you can not upload image. Checkout the <a href="https://www.canner.io/docs/guides-image-upload.html" target="_blank" rel="noreferrer noopener" >storage section</a> to know more</p>} type="warning" />
                )
              }
              <div style={{
                marginBottom: 8
              }}>
                <Com {...this.props}/>
              </div>
            </div>
          }
        </Context.Provider>
      </HOCContext.Provider>
    }
  };
}
