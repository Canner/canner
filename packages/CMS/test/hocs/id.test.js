/**
|--------------------------------------------------
| this is just for copy and past
|--------------------------------------------------
*/


import * as React from 'react';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '../react163Adapter';
import withConnectContext from '../../src/hocs/connectContext';
import {HOCContext as Context} from '../../src/hocs/context';

Enzyme.configure({ adapter: new Adapter() });

describe('withConnectContext', () => {
  let WrapperComponent, props, MockComponent, context,
  mockFetch, mockSubscribe, mockRequest, mockDeploy, mockReset;

  beforeEach(() => {
    MockComponent = function MockComponent() {
      return (<div>Component</div>);
    }
    props = {
      id: undefined,
      keyName: 'posts',
      pattern: 'array',
      routes: ['posts']
    }

    mockFetch = jest.fn();
    mockSubscribe = jest.fn();
    mockRequest = jest.fn();
    mockReset = jest.fn();
    mockDeploy = jest.fn();

    context = {
      componentId: undefined,
      query: undefined,
      fetch: mockFetch,
      subscribe: mockSubscribe,
      request: mockRequest,
      deploy: mockDeploy,
      reset: mockReset
    }

    WrapperComponent = withConnectContext(MockComponent);
  });

  it('should render', () => {
    const wrapper = mount(
      <Context.Provider value={context}>
        <WrapperComponent
          {...props}
        />
      </Context.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should pass props', () => {
    const wrapper = mount(
      <Context.Provider value={context}>
        <WrapperComponent
          {...props}
        />
      </Context.Provider>
    );
    expect(wrapper.find(MockComponent).props()).toHaveProperty('deploy');
    expect(wrapper.find(MockComponent).props()).toHaveProperty('fetch');
    expect(wrapper.find(MockComponent).props()).toHaveProperty('keyName');
    expect(wrapper.find(MockComponent).props()).toHaveProperty('query');
    expect(wrapper.find(MockComponent).props()).toHaveProperty('request');
    expect(wrapper.find(MockComponent).props()).toHaveProperty('reset');
    expect(wrapper.find(MockComponent).props()).toHaveProperty('subscribe');
  });
});