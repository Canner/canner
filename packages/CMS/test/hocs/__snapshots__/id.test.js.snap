// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withConnectContext should render 1`] = `
<ComponentConnectContext
  keyName="posts"
  pattern="array"
  routes={
    Array [
      "posts",
    ]
  }
>
  <MockComponent
    deploy={[MockFunction]}
    fetch={[MockFunction]}
    keyName="posts"
    pattern="array"
    request={[MockFunction]}
    reset={[MockFunction]}
    routes={
      Array [
        "posts",
      ]
    }
    subscribe={[MockFunction]}
  >
    <div>
      Component
    </div>
  </MockComponent>
</ComponentConnectContext>
`;
